ddl/index_merge_tmp_test.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl_test
+
+import (
+	"testing"
+
+	"github.com/pingcap/tidb/ddl"
+	"github.com/pingcap/tidb/ddl/ingest"
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/testkit"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+func TestAddIndexMergeProcess(t *testing.T) {
+	store, dom := testkit.CreateMockStoreAndDomain(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk2 := testkit.NewTestKit(t, store)
+	tk2.MustExec("use test")
+	tk.MustExec("create table t (c1 int primary key, c2 int, c3 int)")
+	tk.MustExec("insert into t values (1, 2, 3), (4, 5, 6);")
+	// Force onCreateIndex use the txn-merge process.
+	ingest.LitInitialized = false
+	tk.MustExec("set @@global.tidb_ddl_enable_fast_reorg = 1;")
+
+	var checkErr error
+	var runDML, backfillDone bool
+	originHook := dom.DDL().GetHook()
+	dom.DDL().SetHook(&ddl.TestDDLCallback{
+		Do: dom,
+		OnJobUpdatedExported: func(job *model.Job) {
+			if !runDML && job.Type == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization {
+				idx := findIdxInfo(dom, "test", "t", "idx")
+				if idx == nil || idx.BackfillState != model.BackfillStateRunning {
+					return
+				}
+				if !backfillDone {
```

Comment:
The state is `BackfillStateReadyToMerge` and `BackfillStateMerging`? This is a little tricky to understand.
And we do nothing for this logic, could we use the need state in line53 directly?

Reply:
No, the state is `BackfillStateRunning`. Because the initialization of `reorgInfo`(reorgInfo.first) takes another round, this hook is called in `BackfillStateRunning` twice.